### PR TITLE
Make MMU and LLEXT compatible

### PR DESCRIPTION
**Pull request by npitre**
For LLEXT to work on MMU systems, we must be able to change memory
attributes for those pages used by LLEXT segments, like making them
executable.

### CODE REVIEW

**Comment by teburd** [comment]
Isn't this already being done if you use the memory partitions and USERMODE=y which enables the memory partition + domain APIs? I guess I'm wondering why this is needed instead of using memory domains and partitions which is a more portable, as I understand things, API.

**Comment by npitre** [comment]
Memory domains/partitions have their own drawbacks, especially if you 
don't need user mode threads to start with. On ARM64, enabling user mode
requires more memory for MMU page tables, etc.

**Comment by teburd** [comment]
> Memory domains/partitions have their own drawbacks

Please elaborate, it was the only portable way of memory permission management when I added it.

If the issue is the API requires enabling usermode, I agree, it should be distinct from usermode.

If there is some other issue with the API then I'd like to hear about it as I'm sure would others so that it can be fixed.

**Comment by npitre** [comment]
> > Memory domains/partitions have their own drawbacks
> 
> Please elaborate, it was the only portable way of memory permission
> management when I added it.

On ARM64 you need a duplicate MMU page table hierarchy for each memory
domain. This is a non-negligeable amount of extra memory. And each of those
domains add overhead to every MMU management operations as global mappings
have to be propagated to every registered domains, necessitating a page
table walk for each of them.

Since a couple years ago I implemented some smartness in the MMU code to
share as many page tables between domains as possible, but the tradeoff for
saving on memory usage is more complex code.

So memory domains are the way to go to support user mode threads. I don't
think the API is bad at all. Simply that there is an inherent cost to it and
it is not the optimal solution for everything.

> If the issue is the API requires enabling usermode, I agree, it should be
> distinct from usermode.

Yes, that's the main issue. If you want user mode threads then you have to
assume the cost that comes with it. No problem there. But that's quite a
significant cost if all you need is LLEXT and no user threads.

Furthermore, as it is, llext_add_domain() always makes a loaded extension
visible to user threads. This means you give permission to execute an
extension to both user and privileged threads together or to no one.

So... if all you want is to make an extension executable from a privileged
thread only, it is a bit overkill to enable user mode support, create and
register a domain, add LLEXT partitions to it, create a new method for
excluding user permissions some time, etc.

Hence this new `k_mem_update_flags()`. It is not universally portable in
the sense that it applies only to MMU equipped systems, but only those
systems require that particular handling anyway. This is lightweight i.e.
the main global page table is modified directly, no need to duplicate the
page table hierarchy, no need to incure the memory domain overhead, etc.
And it doesn't preclude memory domain usage at all if you do need to execute
extension code in user mode threads.

**Review by pillo79 — COMMENTED** [review]
LGTM with one minor comment below :+1:

**Comment by pillo79** [comment]
@npitre side note: as we're changing the access permissions on memory, I think it would be interesting to hear feedback from reviewers on the security aspects of this addition?
Personally I thought this possibility was intentionally missing in the Zephyr APIs, to make "exploits" extremely complicated. But maybe it's simply that nobody has ever needed this before... :slightly_smiling_face:

**Comment by npitre** [comment]
I'm sure security people will tell you that LLEXT should always be 
configured out. This is a very dangerous facility if used unproperly.

But regardless of this, on MMU systems you have to grant execute 
permission to text segments otherwise they're simply unusable. And 
because we can, rodata segments are also marked read-only which is 
arguably better for security than the default read-write.

**Comment by cfriedt** [comment]
LGTM - what do you think @ceolin?

File: subsys/llext/llext_load.c [hunk 1]
```diff
@@ -632,6 +632,8 @@ int do_llext_load(struct llext_loader *ldr, struct llext *ext,
 		goto out;
 	}
 
+	llext_adjust_mmu_permissions(ext);
```
Comment (teburd) [inline subsys/llext/llext_load.c:635]: Is this still needed if CONFIG_USERSPACE=y? A bit unclear to me.

**Comment by npitre** [comment]
> Is this still needed if CONFIG_USERSPACE=y? A bit unclear to me.

Unless `CONFIG_USERSPACE=y` means you need to use every loaded 
extensions in and only in user mode then yes it is still needed.

**Comment by teburd** [comment]
> Is this still needed if CONFIG_USERSPACE=y? A bit unclear to me.
> Unless `CONFIG_USERSPACE=y` means you need to use every loaded extensions in and only in user mode then yes it is still needed.

Why is only usermode required, the tests work with kernel mode extensions today by adding the set of memory partitions defined by each extension to the default (kernel) domain.

**Comment by npitre** [comment]
> Why is only usermode threading required, the tests work with kernel mode
> extensions today by adding the set of memory partitions defined by each
> extension to the default (kernel) domain.

True, my mistake.

However, by applying memory partitions, you open access to extensions for both
user and privileged execution whereas some extensions should be accessible only
to a privileged context.

Given that you do need a mechanism to adjust permissions when
`CONFIG_USERSPACE=n`, and that such mechanism is native to the existing
MMU infrastructure i.e. lower overhead than memory domains, I think it is
better to keep it in all cases and not having to select how privileged
extensions should be set depending on whether or not user thread support
is configured in.

**Comment by teburd** [comment]
> > Why is only usermode threading required, the tests work with kernel mode
> > extensions today by adding the set of memory partitions defined by each
> > extension to the default (kernel) domain.
> 
> True, my mistake.
> 
> However, by applying memory partitions, you open access to extensions for both user and privileged execution whereas some extensions should be accessible only to a privileged context.
> 
> Given that you do need a mechanism to adjust permissions when `CONFIG_USERSPACE=n`, and that such mechanism is native to the existing MMU infrastructure i.e. lower overhead than memory domains, I think it is better to keep it in all cases and not having to select how privileged extensions should be set depending on whether or not user thread support is configured in.

I think we are agreeing here, memory permissions should be settable without all of userspace.

The fundamental question I have is why can't we adjust the domains+partitions API to be useful without userspace instead of adding specialization for MMU here?

**Comment by npitre** [comment]
> The fundamental question I have is why can't we adjust the 
> domains+partitions API to be useful without userspace instead of
> adding specialization for MMU here?

Few reasons come to mind:

- The domains+partitions API in the absence of actual domains+partitions code
  would be a rather roundabout way to achieve the same thing. You'd have to
  create a partition table just to parse it to end up calling the same
  `k_mem_update_flags()` function in some disguised way.

- Pretending that this would be the generic way to do it when in fact only
  MMU systems need this is adding needless complexity to all non-MMU systems
  for no good reasons.

- And similar to this last point, this is a one-off case. I don't see such
  need proliferating. If the abstraction needs more code than the sum of all
  cases where this abstraction is used then this is a bad tradeoff. Let's not
  over-engineer things if not warranted by actual usage.

**Comment by ceolin** [comment]
It does not look great having to deal with the mmu directly here but it seems to me a reasonable approach based on what we have now.

**Review by ceolin — APPROVED** [review]

**Comment by pillo79** [comment]
> > [...] the tests work with kernel mode extensions today by adding the set of memory partitions defined by each extension to the default (kernel) domain.
> 
> [...] by applying memory partitions, you open access to extensions for both user and privileged execution whereas some extensions should be accessible only to a privileged context.

@teburd at the moment we add the domain only to child threads. Would it make sense to test the case you mentioned as well?

@npitre @awojasinski I'm still a little bit confused. Is this new MMU code actually required/tested in the CI suite? AFAICS this PR does not add tests, and the new ones in #77948 are:
* `qemu_cortex_a53: llext.simple.readonly` that should have MMU/MPU disabled 
* `qemu_cortex_a53: llext.simple.readonly_mpu` that defines `USERSPACE=y` and thus domains/mem_parts

I think we need to add an actual third case (`readonly_mmu`?) that leaves MMU enabled but without userspace?

**Comment by npitre** [comment]
qemu_cortex_a53 cannot work at all with MMU disabled.

**Review by teburd — APPROVED** [review]
Begrudgingly accepting.

I spent some time looking into this myself. Pretty clear in reading through the memory domain API... changing mem domains and partitions to be usable outside of userspace is pretty straightforward, a little ifdef moving, to limit things to the default domain.

The terrible part comes when you go to look at how arm64 mmu actually implements this stuff, and what a rabbit hole that is.

It's probably worth revisiting how memory permissions as a whole are being done in llext, and elsewhere. It's useful outside of userspace and its a small travesty that the generic API that should work for MPU + MMU seems to result in... arm64 mmu with its recursive calls and page table copying.

**Review by cfriedt — APPROVED** [review]

**Pull request merged**